### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by thedac**

### CODE REVIEW

File: unit_tests/__init__.py [hunk 1]
```diff
@@ -16,7 +16,15 @@
...
 # Mock out charmhelpers so that we can test without it.
 import charms_openstack.test_mocks  # noqa
 charms_openstack.test_mocks.mock_charmhelpers()
+
+import mock
+import charms
+keystoneauth1 = mock.MagicMock()
+sys.modules['keystoneauth1'] = keystoneauth1
+charms.leadership = mock.MagicMock()
+sys.modules['charms.leadership'] = charms.leadership
```
Comment (pengale) [inline unit_tests/__init__.py:33]: When I've directly inserted mock objects into sys.modules, I've tended to regret it. They don't get reset between tests, so you can wind up polluting one test run with data from another.

I'd recommend instead using the mock.patch start and stop (and maybe reset) commands in your setUp and tearDown methods. (You can also use setUpClass and tearDownClass if you want to wrap the run of every test in a class ... though again, you can run into data pollution issues).

File: unit_tests/test_keystone_saml_mellon_handlers.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        self.remote_id_attribute = "https://samltest.id"
+        self.idp_name = "samltest"
+        self.user_facing_name = "samltest.id"
+        self.keystone_saml_mellon_charm.options.protocol_name = (
+            self.protocol_name)
+        self.keystone_saml_mellon_charm.options.remote_id_attribute = (
+            self.remote_id_attribute)
+        self.keystone_saml_mellon_charm.options.idp_name = self.idp_name
+        self.keystone_saml_mellon_charm.options.user_facing_name = (
+            self.user_facing_name)
```
Comment (pengale) [inline unit_tests/test_keystone_saml_mellon_handlers.py:104]: There are enough mocks here that I'm concerned that this is a functional testing framework masquerading as a unit testing framework. I'd rather see more specific, strategic mocks to test small chunks of code, with functional tests covered by zaza or somet other framework that can actually do functional tests.

Of course, charms are really, really hard to unit test without doing something like this, and unit testing coverage is useful. I'm not sure that this comment is useful beyond just grousing. :-/

**Review by pengale — COMMENTED** [review]
Left some comments. Waiting for tests to run before approving/disapproving.

File: unit_tests/__init__.py (hunk 1 above)
Comment (thedac) [reply unit_tests/__init__.py:33]: Pete, this is a method we have used in many charms. Specifically, it tends to be modules which are not getting tested directly but which may have side effects.

File: unit_tests/test_keystone_saml_mellon_handlers.py (hunk 2 above)
Comment (thedac) [reply unit_tests/test_keystone_saml_mellon_handlers.py:104]: Agreed. For the reactive charms the line is blurry what bits are fully mocked out and what bits are under test. Do you have a specific example or just the general case? Not sure I can fix the general case. But I am happy to look at specific ones.

File: unit_tests/__init__.py (hunk 1 above)
Comment (ajkavanagh) [reply unit_tests/__init__.py:33]: Sorry to chime in here, but:

Mocking in `unit_tests/__init__.py` was *never* meant for actual unit tests, but simply to keep Python happy during import time of the modules *under* tests; i.e. as David says, "... which may have side effects".  Explicitly, with reactive charms, it's to work around the 'self healing' that charm-helpers employs using `import module -> fail -> apt install python-module -> import module`.

Mocks in the `unit_tests/__init__.py` should *never* be assigned return values, and then checked in tests.  That way leads to self harm ... re-mocking is totally appropriate in the unit-test, and it makes it explicit what the unit test and the function-under-test needs.

Comment (pengale) [reply unit_tests/__init__.py:33]: As a fix for self healing, I understand. It's a dangerous solution, but it is a solution. I'd prefer a comment to that effect wherever we do the import, so people know not to use those values in tests.

File: unit_tests/test_keystone_saml_mellon_handlers.py (hunk 2 above)
Comment (pengale) [reply unit_tests/test_keystone_saml_mellon_handlers.py:104]: I was just grousing in general. What's here is fine, and in keeping with what we've done elsewhere.

**Review by pengale — APPROVED** [review]
Comments addressed. LGTM.

**Pull request merged**